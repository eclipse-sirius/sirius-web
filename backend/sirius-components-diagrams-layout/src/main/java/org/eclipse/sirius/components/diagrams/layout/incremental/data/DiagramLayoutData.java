/*******************************************************************************
 * Copyright (c) 2021 THALES GLOBAL SERVICES.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
package org.eclipse.sirius.components.diagrams.layout.incremental.data;

import java.util.List;

import org.eclipse.sirius.components.diagrams.Position;
import org.eclipse.sirius.components.diagrams.Size;

/**
 * A mutable structure to store/update the diagram layout.
 *
 * @author wpiers
 */
public class DiagramLayoutData implements IContainerLayoutData {

    private String id;

    private Position position;

    private Size size;

    private List<NodeLayoutData> nodes;

    private List<EdgeLayoutData> edges;

    @Override
    public String getId() {
        return this.id;
    }

    public void setId(String id) {
        this.id = id;
    }

    @Override
    public Position getPosition() {
        return this.position;
    }

    @Override
    public void setPosition(Position position) {
        this.position = position;
    }

    @Override
    public Size getSize() {
        return this.size;
    }

    @Override
    public void setSize(Size size) {
        this.size = size;
    }

    @Override
    public List<NodeLayoutData> getChildrenNodes() {
        return this.nodes;
    }

    @Override
    public void setChildrenNodes(List<NodeLayoutData> children) {
        this.nodes = children;
    }

    public List<EdgeLayoutData> getEdges() {
        return this.edges;
    }

    public void setEdges(List<EdgeLayoutData> edges) {
        this.edges = edges;
    }

    @Override
    public Position getAbsolutePosition() {
        return this.position;
    }

}
