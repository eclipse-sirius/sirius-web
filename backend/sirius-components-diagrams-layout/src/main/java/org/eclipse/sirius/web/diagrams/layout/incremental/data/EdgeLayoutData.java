/*******************************************************************************
 * Copyright (c) 2021 THALES GLOBAL SERVICES.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
package org.eclipse.sirius.web.diagrams.layout.incremental.data;

import java.util.List;

import org.eclipse.sirius.web.diagrams.Position;
import org.eclipse.sirius.web.diagrams.Ratio;

/**
 * A mutable structure to store/update the edges layout.
 *
 * @author wpiers
 */
public class EdgeLayoutData implements ILayoutData {

    private List<Position> routingPoints;

    private Ratio sourceAnchorRelativePosition;

    private Ratio targetAnchorRelativePosition;

    private NodeLayoutData source;

    private NodeLayoutData target;

    private LabelLayoutData beginLabel;

    private LabelLayoutData centerLabel;

    private LabelLayoutData endLabel;

    private String id;

    @Override
    public String getId() {
        return this.id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public List<Position> getRoutingPoints() {
        return this.routingPoints;
    }

    public void setRoutingPoints(List<Position> routingPoints) {
        this.routingPoints = routingPoints;
    }

    public Ratio getSourceAnchorRelativePosition() {
        return this.sourceAnchorRelativePosition;
    }

    public void setSourceAnchorRelativePosition(Ratio sourceAnchorRelativePosition) {
        this.sourceAnchorRelativePosition = sourceAnchorRelativePosition;
    }

    public Ratio getTargetAnchorRelativePosition() {
        return this.targetAnchorRelativePosition;
    }

    public void setTargetAnchorRelativePosition(Ratio targetAnchorRelativePosition) {
        this.targetAnchorRelativePosition = targetAnchorRelativePosition;
    }

    public NodeLayoutData getSource() {
        return this.source;
    }

    public void setSource(NodeLayoutData source) {
        this.source = source;
    }

    public NodeLayoutData getTarget() {
        return this.target;
    }

    public void setTarget(NodeLayoutData target) {
        this.target = target;
    }

    public LabelLayoutData getBeginLabel() {
        return this.beginLabel;
    }

    public void setBeginLabel(LabelLayoutData beginLabel) {
        this.beginLabel = beginLabel;
    }

    public LabelLayoutData getCenterLabel() {
        return this.centerLabel;
    }

    public void setCenterLabel(LabelLayoutData centerLabel) {
        this.centerLabel = centerLabel;
    }

    public LabelLayoutData getEndLabel() {
        return this.endLabel;
    }

    public void setEndLabel(LabelLayoutData endLabel) {
        this.endLabel = endLabel;
    }
}
