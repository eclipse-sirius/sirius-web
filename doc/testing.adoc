# Testing

This document will describe how we should try to write and maintain tests in Sirius Web.

IMPORTANT: TLDR: Use backend integration tests if you don’t have large and complex pieces of frontend code to test that involve the user interface (diagram interactions, diagram layout, form description editor drag and drop, etc).
They can easily be configured to test very complex use cases involving domains, views, custom services, etc.
They are very fast to execute and are very stable.


## What kind of tests do we have?

We have quite a large scope of tests in Sirius Web with:

- Backend unit tests
- Backend integration tests
- Frontend unit tests
- End to end tests
- Manual tests

## What kind of tests should be written?

The kind of tests that should be written depends on the work that has been done and which should be tested and the goal of the tests.

### Backend unit tests

This kind of tests should be used if you have some very complex algorithm in one specific method.
Given the kind of application that we are developing, this kind of tests should be quite rare.
While we had the backend layout algorithm in the past which represented quite a complex set of algorithms with some geometric computation, this is not really the case anymore.
Since most of the complexity we have to deal with is the integration between components, they are not that useful for us today.

### Frontend unit tests

Just like their backend counterparts, such tests should be written for complex pieces of algorithm on the frontend.
Some of our React component can easily have a complex lifecycle using complex DOM interactions and synchronisation steps with the remote server for example.
They could be tested using this approach but most of the time the experience has shown that this may require setting up some complex mocks of both the DOM and the backend.

Most of the complexity in our frontend code comes from the synchronisation of state between user interactions and the lifecycle of the backend.
As a result, most of the code that needs to be tested in the frontend could just as well be tested by end to end tests.

IMPORTANT: The more mocks are necessary to run unit tests, the further the tests are from the reality.
If you need to mock some parts of the database (with mocks of IRepository for backend tests), network interactions like HTTP requests / WebSocket messages (with mocks of useQuery() or useMutation()) or the DOM (jdom), you probably need integration tests instead.


### Backend integration tests

Those tests are to be considered if most of the logic to be tested is located on the backend and if the frontend part of the tested feature is not significant.

They allow you to test complex state of backend business logic.
It is possible to initialize complex state easily with snapshots of the database or custom services to initialize the editing context with domains or view based representation descriptions.
You can write tests to validate the complex steps of the lifecycle of the representations (create a representation, validate some initial state, trigger some tool, validate the impact of said tool).
We can even contribute test-specific services to mock customization provided by specifiers.

If you don’t have complex pieces of frontend code to test, they are the most efficient way to test Sirius Web applications.


### End to end tests with Cypress

End to end tests are a great way to manipulate the application like the end user.
They give you the ability to test features involving complex user interactions and frontend / backend exchanges.
For example, there is nothing like end to end tests to validate some frontend interactions in diagrams.

While they may seem perfect at first, they also have some sizable drawbacks.
End to end tests are very costly to setup.
Since we are limited to what the user interface and the HTTP API can do, it’s very complex to perform tests with complex initial state.
Testing new features of the view DSL for example, require the creation of a studio project and then an instance project and that’s both cumbersome, long and subject to errors.
 
It is also very difficult to test invalid use cases since both the user interface and the HTTP API prevent end users from creating issues.

End to end tests are also the most complex tests to write and maintain.
Since we have to interact with the user interface like end users, we have to write every single step taken by the end user even if they are not relevant for the feature to be tested.

These tests can easily be flaky since we have to carefully detect user interface states.
Countless tests have been invalid in the past because of a failure to detect that a server response had not been received before trying to perform an action.
For example, to create a representation, you need to wait for the response of the server in the new representation dialog otherwise the list of available representation is empty.


### Manual tests

This is THE perfect test environment since you will end up testing things using real human beings.
But since we are testing with humans, there are a couple things to consider.
Manual tests should be used for two main use cases.

Either to ensure the non regression of use cases that are so complex that we did not succeed in writing automated tests.
For example concurrent tests with multiple users.
Or to leverage the ingenuity of real human beings to find bugs instead of preventing regressions.

#### Using human beings to prevent regressions

Most of the time, preventing regressions involve starting from a specific state, executing a specific set of instructions and validating a specific end result.
This can and should be automated instead since:

- They can be complex to maintain since we won't easily know which tests should be updated after a release (other kinds of tests tends to break quickly after an incompatible change)
- They are boring to execute as an end user
- It’s a shame not to perform such specific test in the continuous integration process
- Human being are very good at missing small and boring details

As such using human for non regression is not really efficient. Let’s be more careful with the contribution of non regression tests when pull results are submitted.

#### Using humans to discover new bugs

Instead it is better to use human beings to use the application to find new bugs.
By giving them some freedom in the way each test should be executed, we can easily end up with different testers using different paths.
In order to discover bugs, the difference in the execution path can sometime be as subtle as clicking on menu items in a different order.
Those manual tests shouldn't be too complex or too long to not frustrate testers. 

## Decision

Here are a couple of directions in which we will move with regards to tests:

- If the vast majority of the behavior being tested is performed by the backend and if the frontend is not very complicated, we will default to backend integration tests
- If a sizable part of the behavior being tested is performed in the frontend, we will default to end to end tests or unit tests
- If a manual test can be automated, we will remove them and use an automated test instead
- End to end tests are very time consuming, we will try to restrain ourselves from writing this kind of test if backend integration tests can do the same job
- Flakky end to end tests are eroding our trust in the build by making everyone fine with red builds.
They will be removed and issues will be created to either restore them if the feature tested had to be tested using an end to end test or to convert them to backend integration tests.
