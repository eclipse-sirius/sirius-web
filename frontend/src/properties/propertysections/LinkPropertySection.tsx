/*******************************************************************************
 * Copyright (c) 2019, 2021 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import IconButton from '@material-ui/core/IconButton';
import Link from '@material-ui/core/Link';
import Snackbar from '@material-ui/core/Snackbar';
import CloseIcon from '@material-ui/icons/Close';
import { useMachine } from '@xstate/react';
import { LinkPropertySectionProps } from 'properties/propertysections/LinkPropertySection.types';
import {
  LinkPropertySectionContext,
  LinkPropertySectionEvent,
  LinkPropertySectionMachine,
  SchemaValue,
} from 'properties/propertysections/LinkPropertySectionMachine';
import React from 'react';

/**
 * Defines the content of a Link property section.
 */
export const LinkPropertySection = ({ editingContextId, formId, widget, subscribers }: LinkPropertySectionProps) => {
  const [{ value: schemaValue, context }, dispatch] = useMachine<LinkPropertySectionContext, LinkPropertySectionEvent>(
    LinkPropertySectionMachine
  );
  const { toast } = schemaValue as SchemaValue;
  const { message } = context;

  return (
    <div>
      <Link id={widget.id} href={widget.url} rel="no-referrer" target="_blank">
        {widget.label}
      </Link>
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'right',
        }}
        open={toast === 'visible'}
        autoHideDuration={3000}
        onClose={() => dispatch({ type: 'HIDE_TOAST' })}
        message={message}
        action={
          <IconButton size="small" aria-label="close" color="inherit" onClick={() => dispatch({ type: 'HIDE_TOAST' })}>
            <CloseIcon fontSize="small" />
          </IconButton>
        }
        data-testid="error"
      />
    </div>
  );
};
