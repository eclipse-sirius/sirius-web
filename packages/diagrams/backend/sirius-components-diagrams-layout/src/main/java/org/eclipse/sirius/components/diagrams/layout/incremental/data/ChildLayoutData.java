/*******************************************************************************
 * Copyright (c) 2023 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
package org.eclipse.sirius.components.diagrams.layout.incremental.data;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

import org.eclipse.sirius.components.annotations.Immutable;
import org.eclipse.sirius.components.diagrams.ILayoutStrategy;
import org.eclipse.sirius.components.diagrams.INodeStyle;
import org.eclipse.sirius.components.diagrams.Position;
import org.eclipse.sirius.components.diagrams.Size;

/**
 * Represents the layout data of a node inside the children area of a node.
 *
 * <p>
 * The child layout data position is <i>not</i> relative to the node parent but to the children area. Once the children
 * area has been calculated, the parent node will go through all of the ChildLayoutData to apply the padding and then
 * check if the position of a node has changed.
 * </p>
 *
 * <p>
 * This is an internal data structure meant to be held by an ChildrenAreaLayoutData.
 * </p>
 *
 * @author gcoutable
 */
@Immutable
public final class ChildLayoutData extends NodeLayoutData {

    private Double paddingTop;

    private Double paddingLeft;

    private ChildLayoutData() {
        // Prevent instantiation
    }

    public static Builder newChildLayoutData(NodeLayoutData childNode) {
        return new Builder(childNode);
    }

    public Double getPaddingTop() {
        return this.paddingTop;
    }

    public Double getPaddingLeft() {
        return this.paddingLeft;
    }

    @Override
    public Position getAbsolutePosition() {
        Position position = super.getAbsolutePosition();
        return Position.at(position.getX() + this.paddingLeft, position.getY() + this.paddingTop);
    }

    /**
     * The builder used to create a children layout data.
     *
     * @author gcoutable
     */
    @SuppressWarnings("checkstyle:HiddenField")
    public static final class Builder {
        private String id;

        private Position position;

        private Size size;

        private boolean userResizable;

        private IContainerLayoutData parent;

        private List<NodeLayoutData> borderNodes;

        private List<NodeLayoutData> nodes;

        private LabelLayoutData label;

        private String nodeType;

        private INodeStyle style;

        private ILayoutStrategy childrenLayoutStrategy;

        private boolean borderNode;

        private boolean changed;

        private boolean pinned;

        private boolean resizedByUser;

        private boolean excludedFromLayoutComputation;

        private Double paddingTop;

        private Double paddingLeft;

        public Builder(NodeLayoutData childNode) {
            this.id = childNode.getId();
            this.size = childNode.getSize();
            this.userResizable = childNode.isUserResizable();
            this.parent = childNode.getParent();
            this.borderNodes = childNode.getBorderNodes();
            this.nodes = childNode.getChildrenNodes();
            this.label = childNode.getLabel();
            this.nodeType = childNode.getNodeType();
            this.style = childNode.getStyle();
            this.childrenLayoutStrategy = childNode.getChildrenLayoutStrategy();
            this.borderNode = childNode.isBorderNode();
            this.changed = childNode.hasChanged();
            this.pinned = childNode.isPinned();
            this.resizedByUser = childNode.isResizedByUser();
            this.excludedFromLayoutComputation = childNode.isExcludedFromLayoutComputation();
        }

        public Builder position(Position position) {
            this.position = Objects.requireNonNull(position);
            return this;
        }

        public Builder paddingTop(Double paddingTop) {
            this.paddingTop = Objects.requireNonNull(paddingTop);
            return this;
        }

        public Builder paddingLeft(Double paddingLeft) {
            this.paddingLeft = Objects.requireNonNull(paddingLeft);
            return this;
        }

        public ChildLayoutData build() {
            ChildLayoutData childLayoutData = new ChildLayoutData();
            childLayoutData.setId(this.id);
            // FIXME: Massive workaround to prevent "unpositioned" node to be "positioned" because their (-1, -1)
            // position whould have become positionInChildArea.
            // This workaround could be removed once the position of node will have become optional globally.
            // There are many _FIXME: Massive workaround (undefined position)_ around the code.
            childLayoutData.setPosition(Optional.ofNullable(this.position).orElse(Position.UNDEFINED));
            childLayoutData.setSize(Objects.requireNonNull(this.size));
            childLayoutData.setUserResizable(this.userResizable);
            childLayoutData.setParent(Objects.requireNonNull(this.parent));
            childLayoutData.setBorderNodes(Objects.requireNonNull(this.borderNodes));
            childLayoutData.setChildrenNodes(Objects.requireNonNull(this.nodes));
            childLayoutData.setLabel(Objects.requireNonNull(this.label));
            childLayoutData.setNodeType(Objects.requireNonNull(this.nodeType));
            childLayoutData.setStyle(Objects.requireNonNull(this.style));
            childLayoutData.setChildrenLayoutStrategy(this.childrenLayoutStrategy);
            childLayoutData.setBorderNode(this.borderNode);
            childLayoutData.setChanged(this.changed);
            childLayoutData.setPinned(this.pinned);
            childLayoutData.setResizedByUser(this.resizedByUser);
            childLayoutData.setExcludedFromLayoutComputation(this.excludedFromLayoutComputation);
            childLayoutData.paddingTop = Objects.requireNonNull(this.paddingTop);
            childLayoutData.paddingLeft = Objects.requireNonNull(this.paddingLeft);
            return childLayoutData;
        }
    }

}
