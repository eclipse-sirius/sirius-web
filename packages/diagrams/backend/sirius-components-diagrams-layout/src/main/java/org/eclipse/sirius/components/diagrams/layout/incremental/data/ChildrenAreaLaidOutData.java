/*******************************************************************************
 * Copyright (c) 2023 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
package org.eclipse.sirius.components.diagrams.layout.incremental.data;

import java.util.Map;
import java.util.Objects;

import org.eclipse.sirius.components.annotations.Immutable;
import org.eclipse.sirius.components.diagrams.Position;
import org.eclipse.sirius.components.diagrams.Size;
import org.eclipse.sirius.components.diagrams.layout.incremental.ILayoutStrategyEngine;
import org.eclipse.sirius.components.diagrams.layout.incremental.INodeIncrementalLayoutEngine;

/**
 * Returns by {@link ILayoutStrategyEngine#getChildrenAreaLayoutData}
 *
 *
 * The data structure used to represent the area of children in a node after it has been laid out.
 *
 * <p>
 * This is an internal data structure meant to be returned by {@link ILayoutStrategyEngine} to the
 * {@link INodeIncrementalLayoutEngine}.
 * </p>
 *
 * @author gcoutable
 */
@Immutable
public final class ChildrenAreaLaidOutData {

    private Position deltaPosition;

    private Size size;

    private Map<String, ChildLayoutData> nodeIdToChildLayoutData;

    private ChildrenAreaLaidOutData() {
        // Prevent instantiation
    }

    public Position getDeltaPosition() {
        return this.deltaPosition;
    }

    public Size getSize() {
        return this.size;
    }

    public Map<String, ChildLayoutData> getNodeIdToChildLayoutData() {
        return this.nodeIdToChildLayoutData;
    }

    public static Builder newChildrenAreaLaidOutData() {
        return new Builder();
    }

    /**
     * The builder used to create a children area laid out data.
     *
     * @author gcoutable
     */
    @SuppressWarnings("checkstyle:HiddenField")
    public static final class Builder {
        private Position deltaPosition;

        private Size size;

        private Map<String, ChildLayoutData> nodeIdToChildLayoutData;

        public Builder deltaPosition(Position deltalPosition) {
            this.deltaPosition = Objects.requireNonNull(deltalPosition);
            return this;
        }

        public Builder size(Size size) {
            this.size = Objects.requireNonNull(size);
            return this;
        }

        public Builder nodeIdToChildLayoutData(Map<String, ChildLayoutData> nodeIdToChildLayoutData) {
            this.nodeIdToChildLayoutData = Objects.requireNonNull(nodeIdToChildLayoutData);
            return this;
        }

        public ChildrenAreaLaidOutData build() {
            ChildrenAreaLaidOutData childrenAreaLayoutData = new ChildrenAreaLaidOutData();
            childrenAreaLayoutData.deltaPosition = Objects.requireNonNull(this.deltaPosition);
            childrenAreaLayoutData.size = Objects.requireNonNull(this.size);
            childrenAreaLayoutData.nodeIdToChildLayoutData = Objects.requireNonNull(this.nodeIdToChildLayoutData);
            return childrenAreaLayoutData;
        }
    }
}
