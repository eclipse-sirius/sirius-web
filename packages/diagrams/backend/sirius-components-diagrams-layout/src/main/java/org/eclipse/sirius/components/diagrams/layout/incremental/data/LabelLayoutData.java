/*******************************************************************************
 * Copyright (c) 2021, 2023 THALES GLOBAL SERVICES.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
package org.eclipse.sirius.components.diagrams.layout.incremental.data;

import org.eclipse.sirius.components.diagrams.Position;
import org.eclipse.sirius.components.diagrams.TextBounds;

/**
 * A mutable structure to store/update the labels layout.
 *
 * @author wpiers
 */
public class LabelLayoutData implements ILayoutData {

    private TextBounds textBounds;

    private Position position;

    private String labelType;

    private String id;

    private boolean isHeader;

    @Override
    public String getId() {
        return this.id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public Position getPosition() {
        return this.position;
    }

    public void setPosition(Position position) {
        this.position = position;
    }

    public String getLabelType() {
        return this.labelType;
    }

    public void setLabelType(String labelType) {
        this.labelType = labelType;
    }

    public TextBounds getTextBounds() {
        return this.textBounds;
    }

    public void setTextBounds(TextBounds textBounds) {
        this.textBounds = textBounds;
    }

    public boolean isHeader() {
        return this.isHeader;
    }

    public void setIsHeader(boolean isHeader) {
        this.isHeader = isHeader;
    }

}
