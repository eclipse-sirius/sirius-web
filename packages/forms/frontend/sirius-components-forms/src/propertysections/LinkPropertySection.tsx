/*******************************************************************************
 * Copyright (c) 2022, 2024 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { getCSSColor } from '@eclipse-sirius/sirius-components-core';
import Link from '@material-ui/core/Link';
import { makeStyles, Theme } from '@material-ui/core/styles';
import { PropertySectionComponent, PropertySectionComponentProps } from '../form/Form.types';
import { GQLLink } from '../form/FormEventFragments.types';
import { getTextDecorationLineValue } from './getTextDecorationLineValue';
import { HelpTooltip } from './HelpTooltip';
import { LinkStyleProps } from './LinkPropertySection.types';

const useStyle = makeStyles<Theme, LinkStyleProps>((theme) => ({
  style: {
    color: ({ color }) => (color ? getCSSColor(color, theme) : null),
    fontSize: ({ fontSize }) => (fontSize ? fontSize : null),
    fontStyle: ({ italic }) => (italic ? 'italic' : null),
    fontWeight: ({ bold }) => (bold ? 'bold' : null),
    textDecorationLine: ({ underline, strikeThrough }) => getTextDecorationLineValue(underline, strikeThrough),
  },
  propertySectionLabel: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
}));

/**
 * Defines the content of a Link property section.
 */
export const LinkPropertySection: PropertySectionComponent<GQLLink> = ({
  editingContextId,
  formId,
  widget,
}: PropertySectionComponentProps<GQLLink>) => {
  const props: LinkStyleProps = {
    color: widget.style?.color ?? null,
    fontSize: widget.style?.fontSize ?? null,
    italic: widget.style?.italic ?? null,
    bold: widget.style?.bold ?? null,
    underline: widget.style?.underline ?? null,
    strikeThrough: widget.style?.strikeThrough ?? null,
  };
  const classes = useStyle(props);

  return (
    <div className={classes.propertySectionLabel}>
      <Link className={classes.style} id={widget.id} href={widget.url} rel="noopener noreferrer" target="_blank">
        {widget.label}
      </Link>
      {widget.hasHelpText ? (
        <HelpTooltip editingContextId={editingContextId} formId={formId} widgetId={widget.id} />
      ) : null}
    </div>
  );
};
