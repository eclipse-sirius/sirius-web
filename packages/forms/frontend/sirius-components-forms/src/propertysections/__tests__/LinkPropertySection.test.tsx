/*******************************************************************************
 * Copyright (c) 2022, 2025 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { MockedProvider } from '@apollo/client/testing';
import { cleanup, render, screen } from '@testing-library/react';
import { afterEach, expect, test } from 'vitest';
import { LinkPropertySection } from '../LinkPropertySection';
import { link } from './LinkPropertySection.data';

afterEach(() => cleanup());

test('should render the link', () => {
  render(
    <MockedProvider>
      <LinkPropertySection editingContextId="editingContextId" formId="formId" widget={link} readOnly={false} />
    </MockedProvider>
  );

  expect(screen.getByRole('link')).toBeDefined();
  expect(screen.getByRole('link').textContent).toBe('Google');
});
