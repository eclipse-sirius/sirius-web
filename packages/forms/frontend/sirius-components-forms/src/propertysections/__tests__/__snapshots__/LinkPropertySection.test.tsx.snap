// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render label widget 1`] = `
<div>
  <div
    class="css-1vrv3vk-propertySectionLabel"
  >
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-17g5t60-style css-1orsvah-MuiTypography-root-MuiLink-root"
      href="the/url/value"
      id="labelId"
      rel="noopener noreferrer"
      target="_blank"
    >
      myLabel
    </a>
  </div>
</div>
`;

exports[`render label widget with help hint 1`] = `
<div>
  <div
    class="css-1vrv3vk-propertySectionLabel"
  >
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-17g5t60-style css-1orsvah-MuiTypography-root-MuiLink-root"
      href="the/url/value"
      id="labelId"
      rel="noopener noreferrer"
      target="_blank"
    >
      myLabel
    </a>
    <svg
      aria-hidden="true"
      aria-label=""
      class="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeMedium css-25to7p-MuiSvgIcon-root"
      data-mui-internal-clone-element="true"
      data-testid="HelpOutlineOutlinedIcon"
      focusable="false"
      style="margin-left: 8px; font-size: 16px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 18h2v-2h-2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8m0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4"
      />
    </svg>
  </div>
</div>
`;

exports[`render label widget with style 1`] = `
<div>
  <div
    class="css-1vrv3vk-propertySectionLabel"
  >
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1mwjg0n-style css-1orsvah-MuiTypography-root-MuiLink-root"
      href="the/url/value"
      id="labelId"
      rel="noopener noreferrer"
      target="_blank"
    >
      myLabel
    </a>
  </div>
</div>
`;
